From e40c071af286598e457c024ed2362995076274f1 Mon Sep 17 00:00:00 2001
From: anonym <anonym@riseup.net>
Date: Wed, 7 Oct 2020 15:50:23 +0200
Subject: [PATCH 2/3] Add pref for whether we accept OAuth2 during
 autoconfiguration.

For many providers JavaScript is required for OAuth2 to work; with it
disabled autoconfiguration then result in a terrible UX (e.g. the web
login fails, has to manually alter the authentication method). Let's
provide a pref that discards OAuth2 configurations so organizations
that disable JavaScript can provide a workaround.
---
 .../accountcreation/content/accountSetup.js   | 66 ++++++++++---------
 .../accountcreation/readFromXML.jsm           | 13 ++++
 comm/mailnews/mailnews.js                     |  2 +
 3 files changed, 51 insertions(+), 30 deletions(-)

diff --git a/comm/mail/components/accountcreation/content/accountSetup.js b/comm/mail/components/accountcreation/content/accountSetup.js
index 8d9852dca30..a9400ce58c4 100644
--- a/comm/mail/components/accountcreation/content/accountSetup.js
+++ b/comm/mail/components/accountcreation/content/accountSetup.js
@@ -1683,36 +1683,42 @@ var gAccountSetup = {
    * @param {AccountConfig} config - The account configuration.
    */
   async adjustOAuth2Visibility(config) {
-    // If the incoming server hostname supports OAuth2, enable it.
-    let iDetails = OAuth2Providers.getHostnameDetails(config.incoming.hostname);
-    document.getElementById("in-authMethod-oauth2").hidden = !iDetails;
-    if (iDetails) {
-      gAccountSetupLogger.debug(
-        `OAuth2 details for incoming server ${config.incoming.hostname} is ${iDetails}`
-      );
-      config.incoming.oauthSettings = {};
-      [
-        config.incoming.oauthSettings.issuer,
-        config.incoming.oauthSettings.scope,
-      ] = iDetails;
-      this._currentConfig.incoming.oauthSettings =
-        config.incoming.oauthSettings;
-    }
-
-    // If the smtp hostname supports OAuth2, enable it.
-    let oDetails = OAuth2Providers.getHostnameDetails(config.outgoing.hostname);
-    document.getElementById("out-authMethod-oauth2").hidden = !oDetails;
-    if (oDetails) {
-      gAccountSetupLogger.debug(
-        `OAuth2 details for outgoing server ${config.outgoing.hostname} is ${oDetails}`
-      );
-      config.outgoing.oauthSettings = {};
-      [
-        config.outgoing.oauthSettings.issuer,
-        config.outgoing.oauthSettings.scope,
-      ] = oDetails;
-      this._currentConfig.outgoing.oauthSettings =
-        config.outgoing.oauthSettings;
+
+
+    if (Services.prefs.getBoolPref("mailnews.auto_config.account_constraints.allow_oauth2")) {
+      // If the incoming server hostname supports OAuth2, enable it.
+      let iDetails = OAuth2Providers.getHostnameDetails(config.incoming.hostname);
+      document.getElementById("in-authMethod-oauth2").hidden = !iDetails;
+      if (iDetails) {
+        gAccountSetupLogger.debug(
+          `OAuth2 details for incoming server ${config.incoming.hostname} is ${iDetails}`
+        );
+        config.incoming.oauthSettings = {};
+        [
+          config.incoming.oauthSettings.issuer,
+          config.incoming.oauthSettings.scope,
+        ] = iDetails;
+        this._currentConfig.incoming.oauthSettings =
+          config.incoming.oauthSettings;
+      }
+    }
+
+    if (Services.prefs.getBoolPref("mailnews.auto_config.account_constraints.allow_oauth2")) {
+      // If the smtp hostname supports OAuth2, enable it.
+      let oDetails = OAuth2Providers.getHostnameDetails(config.outgoing.hostname);
+      document.getElementById("out-authMethod-oauth2").hidden = !oDetails;
+      if (oDetails) {
+        gAccountSetupLogger.debug(
+          `OAuth2 details for outgoing server ${config.outgoing.hostname} is ${oDetails}`
+        );
+        config.outgoing.oauthSettings = {};
+        [
+          config.outgoing.oauthSettings.issuer,
+          config.outgoing.oauthSettings.scope,
+        ] = oDetails;
+        this._currentConfig.outgoing.oauthSettings =
+          config.outgoing.oauthSettings;
+      }
     }
   },
 
diff --git a/comm/mail/components/accountcreation/readFromXML.jsm b/comm/mail/components/accountcreation/readFromXML.jsm
index 1321f85f3a3..ef761f9f2cf 100644
--- a/comm/mail/components/accountcreation/readFromXML.jsm
+++ b/comm/mail/components/accountcreation/readFromXML.jsm
@@ -40,6 +40,8 @@ function readFromXML(clientConfigXML, subSource) {
   function array_or_undef(value) {
     return value === undefined ? [] : value;
   }
+  var allow_oauth2 =
+    Services.prefs.getBoolPref("mailnews.auto_config.account_constraints.allow_oauth2");
   var exception;
   if (
     typeof clientConfigXML != "object" ||
@@ -135,6 +137,12 @@ function readFromXML(clientConfigXML, subSource) {
             NTLM: Ci.nsMsgAuthMethod.NTLM,
             OAuth2: Ci.nsMsgAuthMethod.OAuth2,
           });
+
+          if (!allow_oauth2 && iO.auth == Ci.nsMsgAuthMethod.OAuth2) {
+            iO.auth = null;
+            continue;
+          }
+
           break; // take first that we support
         } catch (e) {
           exception = e;
@@ -278,6 +286,11 @@ function readFromXML(clientConfigXML, subSource) {
             OAuth2: Ci.nsMsgAuthMethod.OAuth2,
           });
 
+          if (!allow_oauth2 && oO.auth == Ci.nsMsgAuthMethod.OAuth2) {
+            oO.auth = null;
+            continue;
+          }
+
           break; // take first that we support
         } catch (e) {
           exception = e;
diff --git a/comm/mailnews/mailnews.js b/comm/mailnews/mailnews.js
index 028b45f98c1..e0d73a3887b 100644
--- a/comm/mailnews/mailnews.js
+++ b/comm/mailnews/mailnews.js
@@ -1081,4 +1081,6 @@ pref("mailnews.auto_config.guess.enabled", true);
 pref("mailnews.auto_config.guess.sslOnly", false);
 pref("mailnews.auto_config.guess.timeout", 10);
+// Whether we allow fetched configurations using OAuth2.
+pref("mailnews.auto_config.account_constraints.allow_oauth2", true);
 pref("dom.xhr.standard_content_type_normalization", false);
 
-- 
2.37.2

